### ISSUE TITLE
Add Lasso

### ISSUE THREAD

**Issue by jakirkham** [dask/dask-ml#101]
Would be good to add [Lasso]( http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html ).

**Comment by TomAugspurger** [comment]
Note: I think that all the pieces should be in place thanks to dask-glm. This should be a matter of translating the scikit-learn API to a linear regression with dask-glm's L1 regularizer.

**Comment by jakirkham** [comment]
Do you have any code snippets that I should look at for trying to do something like this?

**Comment by TomAugspurger** [comment]
I think that

```python
from dask_ml.datasets import make_regression
from dask_glm.regularizers import L1
from dask_glm.estimators import LinearRegression

X, y = make_regression(n_samples=1000, chunks=100)

lr = LinearRegression(regularizer=L1())
lr.fit(X, y)
```

Is basically correct. I haven't looked at the various options for scikit-learn's Lasso.

**Comment by jakirkham** [comment]
Hmm...so when `scikit-learn` implements these sorts of things, they seem to support a vector or matrix for `y`. However it seems that `dask-glm` only supports a vector for `y`. Do you know why that is? Would it be possible to change it? If so, how difficult would that be?

**Comment by TomAugspurger** [comment]
It should certainly be possible, but I'm not sure offhand how much work
it'll be.

**Comment by mrocklin** [comment]
Just checking that by matrix, do you mean ndarray with two dimensions, or
do you mean an np.matrix object?

If the former then is it an array that could be squeezed, or is there
something more complex here with multiple labels?

**Comment by jakirkham** [comment]
Meaning 2-D `ndarray` (though it is a fair question). Should add that `scikit-learn` typically coerces 1-D `ndarray`s into singleton 2-D `ndarray`s when 2-D `ndarray`s are allowed.

Not sure whether squeezing make sense. More likely iterating over the 1-D slices and fitting them independently would make sense, which [appears to be what `scikit-learn` is doing]( https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/linear_model/coordinate_descent.py#L736 ). So this should benefit quite nicely from Distributed.

**Comment by valkmit** [comment]
+1, interested in this as well. The provided code

```python
from dask_ml.datasets import make_regression
from dask_glm.regularizers import L1
from dask_glm.estimators import LinearRegression

X, y = make_regression(n_samples=1000, chunks=100)

lr = LinearRegression(regularizer=L1())
lr.fit(X, y)
```

is missing the ability to set the alpha value - the coefficients seem to point to this not being a proper lasso regression.

**Comment by stsievert** [comment]
Isn't it possible to set the regularization value with the code below?

``` python
from dask_ml.datasets import make_regression
from dask_ml.linear_model import LinearRegression

X, y = make_regression(n_samples=1000, chunks=100)
lr = LinearRegression(regularizer="l1", C=1e-6)
lr.fit(X, y)
assert np.abs(lr.coef_).max() < 1e-3, "C=1e-6 should produce mostly 0 coefs"
```

`C` and `alpha`/`lamduh` control the strength of the regularization (but might be inverses of each other).

**Comment by valkmit** [comment]
> Isn't it possible to set the regularization value with the code below?
> 
> ```python
> from dask_ml.datasets import make_regression
> from dask_ml.linear_model import LinearRegression
> 
> X, y = make_regression(n_samples=1000, chunks=100)
> lr = LinearRegression(regularizer="l1", C=1e-6)
> lr.fit(X, y)
> assert np.abs(lr.coef_).max() < 1e-3, "C=1e-6 should produce mostly 0 coefs"
> ```
> 
> `C` and `alpha`/`lamduh` control the strength of the regularization (but might be inverses of each other).

Indeed this is what I was missing, appreciate the pointer!

**Comment by valkmit** [comment]
Given a small C, the regression does appear to function similarly to Lasso given a small C (implying a large alpha). However, you're also right in that C is inverse of the alpha parameter.

Scikit's documentation says alpha = 1/2C where C is given in other linreg libraries. So an alpha of 0.01 should correspond with a C of 50.

However, with the following code comparing the outputs of both scikit's lasso and Dask's "lasso"

```python
from sklearn.linear_model import Lasso
from dask_ml.datasets import make_regression
from dask_ml.linear_model import LinearRegression

X, y = make_regression(n_samples=1000, chunks=100)
lr = LinearRegression(penalty='l1', C=50, fit_intercept=False)
lr.fit(X, y)

r = Lasso(alpha=0.01, fit_intercept=False)
r.fit(X.compute(), y.compute())

print(lr.coef_)
print(r.coef_)
```

The coefficients for the dask model fit appear unstable. For very small C, they do look the same.

I'm no ML expert - in fact I'm just slapping some code together - but it seems like there's definitely an inverse relationship, just not one that's 1/2C. Which would be fine, except the performance of dask ml at very small C is s**everal times worse than scikit - about 30x worse**, for values of C and alpha that empirically appear to give very similar coefficients.

Furthermore, the coefficients appear to be NaNs when X and y are sourced from a dask dataframe with >1 partition, which seemingly negates any advantage dask would provide in the situation.

Is there something else I am missing here? Or is this performance slowdown to be expected.

**Comment by stsievert** [comment]
> except the performance of dask ml at very small C is several times worse than scikit - about 30x worse, for values of C and alpha that empirically appear to give very similar coefficients.

What do you mean "30× worse"? I'm not sure I'd expect Dask-ML any kind of timing acceleration with a small array.

> C and alpha that empirically appear to give very similar coefficients.

I've verified that C and alpha give very similar coefficients. The two sets of coefficients are very close with relative error, a standard benchmark in optimization:

``` python
# script above
import numpy.linalg as LA
rel_error = LA.norm(lr.coef_ - r.coef_) / LA.norm(r.coef_)
print(rel_error)  # 0.00172; very small. Two vectors have close euclidean distance

print(np.abs(r.coef_).max())  # 89.2532; the scikit-learn coefs are large
print(np.abs(lr.coef_ - r.coef_).mean())  # 0.01543; the mean error is small
print(np.abs(lr.coef_ - r.coef_).max())  # 0.10180; the max error is still pretty large
print(np.median(np.abs(lr.coef_ - r.coef_))) # 0.01077; not as expected (1e-3 or 1e-4 expected), fair given debugging
```
